src/psyclone/psyir/symbols/symboltable.py

```diff
@@ -1077,6 +1084,150 @@ def copy_external_import(self, imported_var, tag=None):
                         " tag '{1}' is already used by another symbol."
                         "".format(imported_var, tag))
 
+    def resolve_imports(self, container_symbols=None, symbol_target=None):
+        ''' Try to resolve deferred and unknown information from imported
+        symbols in this symbol table by searching for their definitions in
+        referred external container. A single symbol to resolve can be
+        specified for a more targeted import.
+
+        :param container_symbols: list of container symbols to search in
+            order to resolve imported symbols. Defaults to all container \
+            symbols in the symbol table.
+        :type container_symbols: list of \
+            :py:class:`psyclone.psyir.symbols.ContainerSymbol`
+        :param symbol_target: If a symbol is given, this method will just \
+            resolve information for the given symbol. Otherwise it will \
+            resolve all possible symbols information. Defaults to None.
+        :type symbol_target: :py:class:`psyclone.psyir.symbols.Symbol` \
+            or NoneType
+
+        :raises SymbolError: if a symbol name clash is found between multiple \
+            imports or an import and a local symbol.
+        :raises TypeError: if the provided container_symbols is not a list of \
+            ContainerSymbols.
+        :raises TypeError: if the provided symbol_target is not a Symbol.
+        :raises KeyError: if a symbol_target has been specified but this has \
+            not been found in any of the searched containers.
+
+        '''
+        if container_symbols is not None:
+            if not isinstance(container_symbols, list):
+                raise TypeError(
+                    f"The resolve_imports container_symbols argument must be a"
+                    f" list but found '{type(container_symbols).__name__}' "
+                    f"instead.")
+            for item in container_symbols:
+                if not isinstance(item, ContainerSymbol):
+                    raise TypeError(
+                        f"The resolve_imports container_symbols argument list "
+                        f"elements must be ContainerSymbols, but found a "
+                        f"'{type(item).__name__}' instead.")
+        else:
+            # If no container_symbol is given, search in all the containers
+            container_symbols = self.containersymbols
+
+        if symbol_target and not isinstance(symbol_target, Symbol):
+            raise TypeError(
+                f"The resolve_imports symbol_target argument must be a Symbol "
+                f"but found '{type(symbol_target).__name__}' instead.")
+
+        for c_symbol in container_symbols:
+            try:
+                external_container = c_symbol.container
+            # pylint: disable=broad-except
+            except Exception:
```

Comment:
When I wrote this I wanted to capture any exception here, my take is:
The IOErrors happen often if the file is not found, but can also be ValueError is a file is found but inside doesn't contain the expected module. This are alright because we don't aim to find all dependencies (and could be named exceptions).

But the reason for a broad except is that then fparser can also (and does in Nemo) fail for some files and this can be NoMatchError, ParseError, FortranSyntaxError, ... We could chose to fail here if something of this happens if you prefer? A compiler would fail for any of these in a dependency. And we can target just specific modules to search for the wildcard import to circumvent these errors. But I wrote this more as a "do whatever it can" and if we find an error in some dependency just ignore this particular dependency and continue. I wouldn't list by name fparser errors because then it means updating/adding fparser errors means we need to update this (and we probably forget until it fails), plus the list of errors is becoming quite long.

I don't think this will contribute to the generic transformation scripts errors messages because this Exception is dropped here and not raised upwards, but of course logging it would be useful to know what had happen in each step of the dependency tree (well we still just do depth 1 for now).

Reply:
OK, and I see you have a TODO to add logging so that's fine.